File: Review-06-CPP-Shapes/Example-7/source/shapeFactory.cpp
// Thomas Kennedy
// CS 330 Fall 2014

#include "shapeFactory.h"

ShapeFactory::ShapePair ShapeFactory::_known_shapes[] = {
    ShapePair( "Triangle" ,             new Triangle()            ),
    ShapePair( "Right Triangle" ,       new RightTriangle()       ),
    ShapePair( "Equilateral Triangle" , new EquilateralTriangle() ),
    ShapePair( "Square",                new Square()              ),
    ShapePair( "Circle",                new Circle()              )
}; // No more empty ShapePair

/**
 *
 */
ShapeFactory::ShapePair::ShapePair()
    : _name(), _model( nullptr )
{
}

/**
 *
 */
ShapeFactory::ShapePair::ShapePair( std::string name, Shape *shape )
    : _name( name ), _model( shape )
{
}

ShapeFactory::ShapePair::~ShapePair()
{
    delete _model;
}

/**
 *
 */
Shape* ShapeFactory::createShape( std::string name )
{
    for(const ShapePair& pair : _known_shapes){
        if( pair._name == name ){
            return pair._model->clone();
        }
    }

    // A shape with the given name could not be found
    return nullptr;
}

/**
 *
 */
bool ShapeFactory::isKnown( std::string name )
{
    for(const ShapePair& pair : _known_shapes){
        if( pair._name == name ){
            return true;
        }
    }

    // The shape with the given name is unknown
    return false;
}

/**
 *
 */
void ShapeFactory::listKnown( std::ostream &outs )
{
    for(const ShapePair& pair : _known_shapes){
        outs << " " << pair._name << "\n"; 
    }
}

/**
 *
 */
int ShapeFactory::numberKnown()
{
    int count = 0;

    for(const ShapePair& pair : _known_shapes){
        count++;
    }

    return count;
}


/**
 *
 */
std::istream& operator>>(std::istream& ins, Shape*& rd)
{
    std::string     name;

    // Read name (key) and create the appropiate Shape
    getline(ins, name, ';');
    rd = ShapeFactory::createShape(name);

    if (rd != nullptr) {
        //do other stuff
        rd->read(ins);
    }
    else {
        // throw away the rest of the line
        getline(ins, name);
    }

    return ins;
}

File: Review-11-Python3-Shapes/Example-4/shapes/ShapeFactory.py
import copy

from shapes.Shape import Shape
from shapes.Circle import Circle
from shapes.Square import Square
from shapes.Triangle import Triangle
from shapes.RightTriangle import RightTriangle
from shapes.EquilateralTriangle import EquilateralTriangle


class ShapeFactory(object):
    """
    The Shape Creating Wizard
    """

    # No more ShapePair
    # Dictionaries are the best
    # kind of data structure

    _known_shapes = {
        "Triangle": (
            Triangle(),
            lambda a, b, c: Triangle(a, b, c)
        ),
        "Right Triangle": (
            RightTriangle(),
            lambda base, height: RightTriangle(base, height)
        ),
        "Equilateral Triangle": (
            EquilateralTriangle(),
            lambda side: EquilateralTriangle(side)
        ),
        "Square": (
            Square(),
            lambda side: Square(side)
        ),
        "Circle": (
            Circle(),
            lambda radius: Circle(radius)
        )
    }  # _Dictionary_ of known shapes

    @staticmethod
    def create(name):
        """
        Create a Shape

        :param: name the shape to be created

        :return: A shape with the specified name
           or null if no matching shape is found
        """

        if name in ShapeFactory._known_shapes:
            return copy.deepcopy(
                ShapeFactory._known_shapes[name][0]
            )

        return None

    @staticmethod
    def createFromDictionary(name, values):
        """
        Create a Shape

        :param name: the shape to be created
        :param attributes: dictionary of values corresponding
            to the data needed to inialize a shape

        :return: A shape with the specified name
           or null if no matching shape is found
        """

        if name in ShapeFactory._known_shapes:
            return (
                ShapeFactory._known_shapes[name][1](**values)
            )

        return None

    @staticmethod
    def isKnown(name):
        """
        Determine whether a given shape is known

        :param: name the shape for which to query
        """

        return (name in ShapeFactory._known_shapes)

    @staticmethod
    def listKnown():
        """
        Print a list of known Shapes
        """
        return "\n".join(
            ["  " + name for name in ShapeFactory._known_shapes.keys()])

    @staticmethod
    def numberKnown():
        """
        Determine the number of known Shapes
        """

        return len(ShapeFactory._known_shapes)
